lib/Core/Memory.h

```diff
@@ -248,6 +255,103 @@ class ObjectState {
 
   ArrayCache *getArrayCache() const;
 };
+
+/// \brief A subclass of ObjectState used for tracking the state of a
+/// persistent MemoryObject.
+///
+/// Tracks updates to a symbolic Array of bools that represents the
+/// dirtiness/persistedness of each cache line that the MemoryObject spans.
+class PersistentState : public ObjectState {
+  private:
+    /// Both of these UpdateLists share one underlying symbolic bool Array.
+    /// The Array has one entry per cache line: 1 if persisted, 0 if not.
+    /// Each write and each cache line flush are tracked as updates to this array.
+    ///
+    /// These two update lists diverge during a persistence epoch (epochs are
+    /// delimited by memory fences).
+    /// Until the next sfence, cacheLineUpdates contains only the writes made
+    /// during the current epoch, while pendingCacheLineUpdates contains both
+    /// the writes and flushes.
+    ///
+    /// Queries about the persistence of cache lines are formed using
+    /// cacheLineUpdates rather than pendingCacheLineUpdates.
+    /// Thus, any persistence queries made during the middle of an epoch will
+    /// not falsely report that writes have been persisted.
+    ///
+    /// When the epoch ends, the pending flushes are added to cacheLineUpdates,
+    /// preserving their original ordering WRT the writes.
```

Comment:
And the writes too, right?

Reply:
Yes. From one point of view, the writes were already there, but we were keeping the flushes out until the fence

Reply:
Gotcha. Comment was just a little vague.